File: python/traversal_engine.py
# =============================================================================
# vapor-idx — traversal_engine.py
# =============================================================================

from __future__ import annotations
from collections import deque
from .types import (
    TraversalOptions, TraversalResult, TraversalEntry,
    PathOptions, VaporError,
)
from .record_store import RecordStore
from .relationship_store import RelationshipStore
from .query_engine import QueryEngine


class TraversalEngine:
    def __init__(
        self,
        records: RecordStore,
        relationships: RelationshipStore,
        query: QueryEngine,
    ) -> None:
        self._records       = records
        self._relationships = relationships
        self._query         = query

    def traverse(self, options: TraversalOptions) -> TraversalResult:
        from_id      = options.from_id
        rel_type     = options.relationship
        direction    = options.direction
        max_depth    = options.depth
        filt         = options.filter

        if not self._records.has(from_id):
            raise VaporError(f'Traversal start record "{from_id}" does not exist.')

        visited: set[str]          = {from_id}
        entries: list[TraversalEntry] = []
        result_records               = []

        # BFS: (current_id, depth, via)
        queue: deque[tuple[str, int, list[str]]] = deque([(from_id, 0, [])])

        while queue:
            current_id, current_depth, via = queue.popleft()
            if current_depth >= max_depth:
                continue

            neighbours = self._relationships.get_neighbour_ids(current_id, rel_type, direction)
            for neighbour_id in neighbours:
                if neighbour_id in visited:
                    continue
                visited.add(neighbour_id)

                record = self._records.get(neighbour_id)
                if record is None:
                    continue

                if filt:
                    match = self._query.query(filt)
                    if not any(r.id == neighbour_id for r in match.records):
                        continue

                entry = TraversalEntry(record=record, depth=current_depth + 1, via=list(via) + [current_id])
                entries.append(entry)
                result_records.append(record)

                if current_depth + 1 < max_depth:
                    queue.append((neighbour_id, current_depth + 1, list(via) + [current_id]))

        return TraversalResult(records=result_records, entries=entries)

    def find_path(self, options: PathOptions) -> list[str] | None:
        from_id  = options.from_id
        to_id    = options.to_id
        rel_type = options.relationship
        max_d    = options.max_depth

        if not self._records.has(from_id):
            raise VaporError(f'Path start "{from_id}" does not exist.')
        if not self._records.has(to_id):
            raise VaporError(f'Path end "{to_id}" does not exist.')
        if from_id == to_id:
            return [from_id]

        visited: set[str]               = {from_id}
        queue:   deque[tuple[str, list[str]]] = deque([(from_id, [from_id])])

        while queue:
            current_id, path = queue.popleft()
            if len(path) - 1 >= max_d:
                continue
            edges = self._relationships.get_edges_for_record(current_id, rel_type, "both")
            for edge in edges:
                neighbour_id = edge.target_id if edge.source_id == current_id else edge.source_id
                if neighbour_id in visited:
                    continue
                visited.add(neighbour_id)
                new_path = path + [neighbour_id]
                if neighbour_id == to_id:
                    return new_path
                queue.append((neighbour_id, new_path))
        return None

File: rust/src/traversal_engine.rs
// =============================================================================
// vapor-idx — traversal_engine.rs
// BFS-based graph traversal with shortest-path support.
// =============================================================================

use crate::types::*;
use crate::record_store::RecordStore;
use crate::relationship_store::RelationshipStore;
use crate::query_engine::QueryEngine;
use std::collections::{HashSet, VecDeque};

pub struct TraversalEngine<'a> {
    records:       &'a RecordStore,
    relationships: &'a RelationshipStore,
    query:         &'a QueryEngine<'a>,
}

impl<'a> TraversalEngine<'a> {
    pub fn new(
        records:       &'a RecordStore,
        relationships: &'a RelationshipStore,
        query:         &'a QueryEngine<'a>,
    ) -> Self {
        Self { records, relationships, query }
    }

    pub fn traverse(&self, options: &TraversalOptions) -> VaporResult<TraversalResult> {
        if !self.records.has(&options.from) {
            return Err(VaporError::NotFound(format!("Start record \"{}\" not found.", options.from)));
        }

        let direction = match &options.direction {
            TraversalDirection::Outgoing => "outgoing",
            TraversalDirection::Incoming => "incoming",
            TraversalDirection::Both     => "both",
        };

        let mut visited:  HashSet<String>   = HashSet::from([options.from.clone()]);
        let mut entries:  Vec<TraversalEntry> = Vec::new();
        let mut records:  Vec<VaporRecord>    = Vec::new();

        // BFS queue: (current_id, depth, via_path)
        let mut queue: VecDeque<(String, usize, Vec<String>)> =
            VecDeque::from([(options.from.clone(), 0, vec![])]);

        while let Some((current_id, current_depth, via)) = queue.pop_front() {
            if current_depth >= options.depth { continue; }

            let neighbours = self.relationships.get_neighbour_ids(
                &current_id, &options.relationship, direction
            );

            for neighbour_id in neighbours {
                if visited.contains(&neighbour_id) { continue; }
                visited.insert(neighbour_id.clone());

                let Some(record) = self.records.get(&neighbour_id) else { continue };

                // Optional filter
                if let Some(filter) = &options.filter {
                    let result = self.query.query(filter)?;
                    if !result.records.iter().any(|r| r.id == neighbour_id) { continue; }
                }

                let mut new_via = via.clone();
                new_via.push(current_id.clone());

                entries.push(TraversalEntry {
                    record: record.clone(),
                    depth:  current_depth + 1,
                    via:    new_via.clone(),
                });
                records.push(record.clone());

                if current_depth + 1 < options.depth {
                    queue.push_back((neighbour_id, current_depth + 1, new_via));
                }
            }
        }

        Ok(TraversalResult { records, entries })
    }

    pub fn find_path(&self, options: &PathOptions) -> VaporResult<Option<Vec<String>>> {
        if !self.records.has(&options.from) {
            return Err(VaporError::NotFound(format!("Start \"{}\" not found.", options.from)));
        }
        if !self.records.has(&options.to) {
            return Err(VaporError::NotFound(format!("End \"{}\" not found.", options.to)));
        }
        if options.from == options.to {
            return Ok(Some(vec![options.from.clone()]));
        }

        let mut visited: HashSet<String> = HashSet::from([options.from.clone()]);
        let mut queue: VecDeque<Vec<String>> = VecDeque::from([vec![options.from.clone()]]);

        while let Some(path) = queue.pop_front() {
            if path.len() - 1 >= options.max_depth { continue; }

            let current_id = path.last().unwrap();
            let edges = self.relationships.get_edges_for_record(
                current_id,
                options.relationship.as_deref(),
                "both",
            );

            for edge in edges {
                let neighbour_id = if edge.source_id == *current_id {
                    &edge.target_id
                } else {
                    &edge.source_id
                };

                if visited.contains(neighbour_id) { continue; }
                visited.insert(neighbour_id.clone());

                let mut new_path = path.clone();
                new_path.push(neighbour_id.clone());

                if *neighbour_id == options.to {
                    return Ok(Some(new_path));
                }

                queue.push_back(new_path);
            }
        }

        Ok(None)
    }
}
